lib/diskoverflow/map.go

```diff
@@ -0,0 +1,315 @@
+// method is called in case spilling happened.
+func (o *omap) Close() {
+	o.commonMap.Close()
+}
+
+func (o *omap) released() {
+	o.iterating = false
+}
+
+type MapIterator interface {
+	Iterator
+	Key() string
+}
+
+func (o *omap) NewIterator() MapIterator {
+	if o.iterating {
+		panic(concurrencyMsg)
+	}
+	return o.newIterator(o)
+}
+
+type memoryMap struct {
+	values map[string]Value
+	bytes  int
+}
+
+func (o *memoryMap) set(k string, v Value) {
+	o.values[k] = v
+	o.bytes += v.ProtoSize()
+}
+
+func (o *memoryMap) Bytes() int {
+	return o.bytes
+}
+
+func (o *memoryMap) Close() {
+	o.values = nil
+}
+
+func (o *memoryMap) Get(k string, v Value) bool {
+	nv, ok := o.values[k]
+	if !ok {
+		return false
+	}
+	copyValue(v, nv)
+	return true
+}
+
+func (o *memoryMap) Items() int {
+	return len(o.values)
+}
+
+func (o *memoryMap) Pop(k string, v Value) bool {
+	ok := o.Get(k, v)
+	if !ok {
+		return false
+	}
+	delete(o.values, k)
+	o.bytes -= v.ProtoSize()
+	return true
+}
+
+func (o *memoryMap) Delete(k string) {
+	v, ok := o.values[k]
+	if !ok {
+		return
+	}
+	delete(o.values, k)
+	o.bytes -= v.ProtoSize()
+}
+
+type iteratorValue struct {
+	k string
+	v Value
+}
+
+type memMapIterator struct {
+	values  map[string]Value
+	ch      chan iteratorValue
+	stop    chan struct{}
+	current iteratorValue
+	parent  iteratorParent
+}
+
+func (o *memoryMap) newIterator(p iteratorParent) MapIterator {
+	i := &memMapIterator{
+		ch:     make(chan iteratorValue),
+		stop:   make(chan struct{}),
+		parent: p,
+		values: o.values,
+	}
+	go i.iterate()
+	return i
+}
+
+func (i *memMapIterator) iterate() {
+	for k, v := range i.values {
+		select {
+		case <-i.stop:
+			break
+		case i.ch <- iteratorValue{k, v}:
+		}
+	}
+	close(i.ch)
+}
+
+func (i *memMapIterator) Key() string {
+	return i.current.k
+}
+
+func (i *memMapIterator) Value(v Value) {
+	copyValue(v, i.current.v)
+}
+
+func (i *memMapIterator) Next() bool {
+	var ok bool
+	i.current, ok = <-i.ch
+	return ok
+}
+
+func (i *memMapIterator) Release() {
+	close(i.stop)
+	i.parent.released()
+}
+
+type diskMap struct {
+	db    *leveldb.DB
+	bytes int
+	dir   string
+	len   int
+}
+
+func newDiskMap(location string) *diskMap {
+	// Use a temporary database directory.
+	tmp, err := ioutil.TempDir(location, "overflow-")
+	if err != nil {
+		panic("creating temporary directory: " + err.Error())
+	}
+	db, err := leveldb.OpenFile(tmp, &opt.Options{
+		OpenFilesCacheCapacity: 10,
+		WriteBuffer:            512 << 10,
+	})
+	if err != nil {
+		panic("creating temporary database: " + err.Error())
+	}
+	return &diskMap{
+		db:  db,
+		dir: tmp,
+	}
+}
+
+func (o *diskMap) set(k string, v Value) {
+	o.addBytes([]byte(k), v)
+	o.bytes += v.ProtoSize()
+}
+
+func (o *diskMap) addBytes(k []byte, v Value) {
+	d, err := v.Marshal()
+	errPanic(err)
+	if err := o.db.Put(k, d, nil); err != nil {
+		panic("writing to temporary database: " + err.Error())
+	}
+	o.len++
+}
```

Comment:
Potentially want to do this in a batch when spilling.
Perhap doing batching in general is not a bad idea.

Reply:
Not sure whether that would have any effect: As far as I understood batching isn't about performance, but about atomicity of changes. First stage of goleveldb is in memory, so in terms of disk IO batching happens under the hood. In the docs for `db.Write` it is mentioned that multiple `Put`s, `Writes` of batches, ... could be merged underneath when called concurrently - not sure how big the window is, i.e. if it may apply to the non-concurrent loop when spilling.

Reply:
I think batching prevents new level creation half way through the write. Why would we batch index updates every 250 items otherwise?

Reply:
That the (unanswered) question I already brought up here: https://github.com/syncthing/syncthing/issues/5531#issuecomment-463592922  
I'd expect level creation to happen when the (internal) memdb is flushed to disk. As we do stuff sequentially, there might still be a bit of saved overhead in batching stuff up, however I suspect it doesn't as it batches stuff internally. I didn't check the code though, so that's just my interpretation of documentation and some comments on goleveldbs issue tracker.